library/psa_crypto.c

```diff
@@ -535,6 +535,53 @@ psa_status_t psa_allocate_buffer_to_slot( psa_key_slot_t *slot,
     return( PSA_SUCCESS );
 }
 
+/** Calculate the expected size of a key buffer, based on its attributes
+ *
+ * This function provides a way to get the expected size for storing a key with
+ * the given attributes. This will be the size of the export representation for
+ * cleartext keys, and a driver-defined size for keys stored by opaque drivers.
+ *
+ * \param[in] attributes        The key attribute structure of the key to store.
+ * \param[in] key_bits          The actual key bits used to deduce the size, given
+ *                              that this could vary from the core key_bits in the
+ *                              key attributes.
+ * \param[inout] storage_size   On success, a byte size large enough to contain
+ *                              the declared key. Prepopulated by the caller
+ *                              with the amount of input data in case key_bits
+ *                              is unknown.
+ *
+ * \retval #PSA_SUCCESS
+ * \retval #PSA_ERROR_INVALID_ARGUMENT
+ * \retval #PSA_ERROR_NOT_SUPPORTED
+ */
+static psa_status_t psa_calculate_key_storage_size( const psa_key_attributes_t *attributes,
+                                                    size_t key_bits,
+                                                    size_t* storage_size )
+{
+    if( psa_key_lifetime_is_external( attributes->core.lifetime ) )
+    {
+#if defined(MBEDTLS_PSA_CRYPTO_DRIVERS)
+        return psa_driver_wrapper_get_key_buffer_size( attributes, key_bits, storage_size );
+#else
+        return( PSA_ERROR_INVALID_ARGUMENT );
+#endif
+    }
+    else
+    {
+        if( key_bits != 0 )
+        {
+            size_t expected_size = PSA_EXPORT_KEY_OUTPUT_SIZE( attributes->core.type, key_bits );
```

Comment:
Yes this part is not specific to opaque keys , but do we agree that this sizing is required ?

Reply:
I agree that in `psa_import_key()` just relying on `data_length` to determine the key buffer size is not ideal and will need to be improved but I am not convinced by the calculation here. If a transparent key is imported for the sake of a driver accelerator the key export output size may be not the proper size as the accelerator may prefer another key representation for its operations.

Reply:
The key export size is not the most appropriate size, but could it be considered as the max size that is required for any given key types' representation ? If it could, this scales till we have a more detailed sizing operation that sizes for each key type by parsing the key. What do you think ?

Reply:
I was of the impression that the export format was rather on "min size" side but I can be wrong there. Anyway as you see there are quite a lot of uncertainties here thus I would prefer just to leave that for another time.

Reply:
Actually if you look at the patch, the test drivers' sizing is  based on the export notation, assuming that to be the max given a key type. So we do need closure on that (Is the export notation the max size for a given key type).
This was the point that was left open earlier on the review with @gilles-peskine-arm.

While we can keep fixing the transparent key sizing for a later time, I'd like to understand if the export notation is the minimum or max representation of a key type.

Reply:
> Actually if you look at the patch, the test drivers' sizing is based on the export notation, assuming that to be the max given a key type. So we do need closure on that (Is the export notation the max size for a given key type).

I am not sure what you are referring to here, could you elaborate?

Reply:
Bumped a comment in second commit to reference this.
To simply put it I'd say " Is the export notation the 'max size' or 'min size' of a representation" is something that needs to be clarified if we want to go any further on this PR.